jetty-security/src/main/java/org/eclipse/jetty/security/SecurityHandler.java

```diff
@@ -572,6 +572,11 @@ else if (authentication instanceof Authentication.Deferred)
                             authenticator.secureResponse(request, response, isAuthMandatory, null);
                     }
                 }
```

Comment:
Should we also check for mandatory auth here, as I don't think we should defer mandatory auth.

Reply:
I tried enforcing `isAuthMandatory` for `DeferredAuthentication` as well but it broke things, apparently some of the authenticators are using DeferredAuthentication to bypass the security constraints on purpose, for example FORM auth does this to allow access to the form page.

Reply:
Deferred auth is the exact opposite of mandatory. If a request has credentials but his a none mandatory URI, then we used deferred auth to avoid the cost of checking the credentials unless somebody calls a auth method

Reply:
I understand that. 

I am just saying that if we change the code to not allow deferred auth for isMandatory==true we will break how some of our existing Authenticators work.

Reply:
I'm not saying that.  The spirit of this PR is to make the code here more robust in the face of unexpected returns from `validateRequest`, so  I'm saying that if `validateRequest` above somehow returned a `Deferred` auth then in this section we should check for `isMandatory==true` and if so resolve the authentication.

Actually we should probably do that above, so the results of that authentication can be fed into this if block.

Reply:
OK - I now understand that our own authenticators are returning deferred auth to indicate a temporary OK for login pages and similar.   So I think this needs to be cleaned up... but not in this PR.